daemon/logger/fluentd/fluentd.go

```diff
@@ -172,29 +185,65 @@ func ValidateLogOpt(cfg map[string]string) error {
 		}
 	}
 
-	if _, _, err := parseAddress(cfg["fluentd-address"]); err != nil {
+	if _, err := parseAddress(cfg["fluentd-address"]); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func parseAddress(address string) (string, int, error) {
+func parseAddress(address string) (*location, error) {
 	if address == "" {
-		return defaultHost, defaultPort, nil
+		return &location{
+			protocol: defaultProtocol,
+			host:     defaultHost,
+			port:     defaultPort,
+			path:     "",
+		}, nil
+	}
+
+	protocol := defaultProtocol
+	givenAddress := address
+	if urlutil.IsTransportURL(address) {
+		url, err := url.Parse(address)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid fluentd-address %s", givenAddress)
+		}
+		// unix and unixgram socket
+		if url.Scheme == "unix" || url.Scheme == "unixgram" {
+			return &location{
+				protocol: url.Scheme,
+				host:     "",
+				port:     0,
+				path:     url.Path,
+			}, nil
+		}
+		// tcp|udp
+		protocol = url.Scheme
+		address = url.Host
 	}
 
 	host, port, err := net.SplitHostPort(address)
 	if err != nil {
 		if !strings.Contains(err.Error(), "missing port in address") {
-			return "", 0, fmt.Errorf("invalid fluentd-address %s: %s", address, err)
+			return nil, errors.Wrapf(err, "invalid fluentd-address %s", givenAddress)
 		}
-		return host, defaultPort, nil
+		return &location{
+			protocol: protocol,
+			host:     host,
+			port:     defaultPort,
+			path:     "",
+		}, nil
 	}
 
 	portnum, err := strconv.Atoi(port)
 	if err != nil {
-		return "", 0, fmt.Errorf("invalid fluentd-address %s: %s", address, err)
+		return nil, errors.Wrapf(err, "invalid fluentd-address %s", givenAddress)
 	}
-	return host, portnum, nil
+	return &location{
```

Comment:
I suggested to change, because the initial implementation was having 5 output variables https://github.com/docker/docker/pull/26088#discussion_r84602329, but perhaps we can reduce that some other ways, agreed